relm4/src/sender.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright 2022 System76 <info@system76.com>
+// SPDX-License-Identifier: MIT or Apache-2.0
+
+//! Contains various flavors of channels to send messages between components and workers.
+
+use crate::factory::FactoryComponent;
+use crate::{Component, Sender, ShutdownReceiver};
+use std::future::Future;
+
+/// Contains senders to send and receive messages from a [`Component`].
+#[derive(Debug)]
+pub struct ComponentSender<C: Component> {
+    /// Emits inputs to the component.
+    pub input: Sender<C::Input>,
+    /// Emits outputs from the component.
+    pub output: Sender<C::Output>,
+    command: Sender<C::CommandOutput>,
+    shutdown: ShutdownReceiver,
+}
+
+impl<C: Component> ComponentSender<C> {
+    pub(crate) fn new(
+        input_tx: Sender<C::Input>,
+        output_tx: Sender<C::Output>,
+        command_tx: Sender<C::CommandOutput>,
+        shutdown_tx: ShutdownReceiver,
+    ) -> Self {
+        ComponentSender {
+            input: input_tx,
+            output: output_tx,
+            command: command_tx,
+            shutdown: shutdown_tx,
+        }
+    }
+
+    /// Spawns a command.
+    /// You can bind the the command to the lifetime of the component
+    /// by using a [`ShutdownReceiver`].
+    pub fn command<Cmd, Fut>(&self, cmd: Cmd)
+    where
+        Cmd: FnOnce(Sender<C::CommandOutput>, ShutdownReceiver) -> Fut + Send + 'static,
+        Fut: Future<Output = ()> + Send,
+    {
+        let recipient = self.shutdown.clone();
+        let sender = self.command.clone();
+        crate::spawn(async move {
+            cmd(sender, recipient).await;
+        });
+    }
+
+    /// Spawns a future that will be dropped as soon as the component is shut down.
+    ///
+    /// Essentially, this is a simpler version of [`Self::command()`].
+    pub fn oneshot_command<Fut>(&self, future: Fut)
+    where
+        Fut: Future<Output = C::CommandOutput> + Send + 'static,
+    {
+        // Async closures would be awesome here...
+        self.command(move |out, shutdown| {
+            shutdown
+                .register(async move { out.send(future.await) })
+                .drop_on_shutdown()
+        });
+    }
+
+    /// Emit an input to the component.
+    pub fn input(&self, message: C::Input) {
+        self.input.send(message)
+    }
+
+    /// Equivalent to `&self.input`.
+    #[must_use]
+    pub fn input_sender(&self) -> &Sender<C::Input> {
+        &self.input
+    }
```

Comment:
I'd say we either make this field public xor have a getter method. I think it'd usually be better to hide the internal fields...

Reply:
Agreed. If we're fine making another large breaking change before 0.5, then I'd propose just removing `input` and `output` in favor of `input_sender` and `output_sender`, though this will likely involve updating almost every `Component` impl out there.

Current status:

| Retrieve input sender | Retrieve output sender | Send input | Send output |
| ----------------------------- | -------------------------------| ----------- | --------|
| `.input` or `.input_sender()` | `.output` or `.output_sender()` | `.input(msg)` | `.output(msg)` |

Proposed:

| Retrieve input sender | Retrieve output sender | Send input | Send output |
| ----------------------------- | -------------------------------| ----------- | --------|
| `.input_sender()` | `.output_sender()` | `.input(msg)` | `.output(msg)` |

Removing the `input` and `output` fields will also allow us to bring back `ComponentSenderInner` to share implementation between the two.

Reply:
The proposed changes sound good to me. Also I don't think the senders are often used directly, but by sending messages with `.input()` for example. So I don't think it will affect a lot of existing 0.5 projects. Yet in any case it's not a very big breaking change either way.

Reply:
Implemented!